src/main/scala/io/iohk/ethereum/domain/BlockchainReader.scala

```diff
@@ -113,6 +111,38 @@ class BlockchainReader(
   def genesisBlock: Block =
     getBlockByNumber(0).get
 
+  /** Returns a block inside this branch based on its number */
+  def getBlockByNumber(branch: Branch, number: BigInt): Option[Block] = branch match {
+    case BestBranch(_, tipBlockNumber) =>
+      if (tipBlockNumber >= number && number >= 0) {
+        for {
+          hash <- getHashByBlockNumber(number)
+          block <- getBlockByHash(hash)
+        } yield block
+      } else None
+    case EmptyBranch => None
+  }
+
+  /** Returns a block hash for the block at the given height if any */
+  def getHashByBlockNumber(branch: Branch, number: BigInt): Option[ByteString] = branch match {
+    case BestBranch(_, tipBlockNumber) =>
+      if (tipBlockNumber >= number && number >= 0) {
+        blockNumberMappingStorage.get(number)
+      } else None
+
+    case EmptyBranch => None
+  }
+
+  /** Checks if given block hash is in this chain. (i.e. is an ancestor of the tip block) */
+  def isInChain(branch: Branch, hash: ByteString): Boolean = branch match {
+    case BestBranch(_, tipBlockNumber) =>
+      (for {
+        header <- getBlockHeaderByHash(hash) if header.number <= tipBlockNumber
+        hash <- getHashByBlockNumber(branch, header.number)
```

Comment:
You are shadowing `hash` here; perhaps you can specify the names a bit more?

Also, this relies on using the "hash by block number" index as an authority of "this chain". How do you envision that to work in the new architecture?

Reply:
Good catch. Fixed

For now we have to keep the index updated indeed. Right now `isInChain` is used by `BranchResolution.resolveBranch` to check if a sequence of blocks has a parent inside the canonical chain (i.e. if the sequence is rooted). In the end, we would like for mantis to be able to synchronize without maintaining that index. It would still be there to help json rpc calls, but mantis would not have to wait for it to be rebuilt to continue its work.

That said, in the new design, we won't really care that a parent is part of the chain, but we will care that a block is rooted so that we can compare its weight to the current best block.

If we want to do that, I think it depend on how we store incoming blocks: 
 - if only rooted block are stored (i.e. we store block once they form a branch), then if it is in storage, all is good and we can compare it with the best branch.
 - if we store every block it gets trickier. I think in that case we will have to add some block metadata which tells us if a block is rooted or not so that we can easily tell "the parent is rooted so this block is also rooted", probably with different types of blocks.